allennlp/training/trainer.py

```diff
@@ -992,11 +992,20 @@ def from_params(cls,
         grad_norm = params.pop_float("grad_norm", None)
         grad_clipping = params.pop_float("grad_clipping", None)
         lr_scheduler_params = params.pop("learning_rate_scheduler", None)
+        optimizer_params = params.pop("optimizer")
+        if isinstance(optimizer_params, str):
+            optimizer_params = Params({"type": optimizer_params})
+        if "parameter_groups" in optimizer_params:
```

Comment:
I really don't like having to do this here.  I don't really understand well enough what this is accomplishing, so I don't know if it's feasible to modify the logic inside of your LR scheduler to make it so you can remove this.  But if there's a way to do it, I think it would be much better than doing this.  This is a pretty obscure dependency that's introduced between a LR scheduler and an optimizer; it's not clear at all from looking at this code that the point is to make your LR scheduler work.  It'd be really easy for a future PR to unwittingly break your scheduler by modifying these lines.

If there's no way to modify the logic in the LR scheduler to get the behavior you want, then it looks like you can just require a particular specification of the optimizer parameters in the config file, that you can check and enforce in the LR scheduler - the parameters after this modification are still something that could have been specified in JSON.  That's better to me than breaking the encapsulation of the optimizer and the LR scheduler by having these lines here.

Reply:
I agree. As I mentioned above, if we don't want to do this, then the parameter groups need to be specified in the optimizer, which is fine. I did not find a clear way to split the single list of parameters in the `Optimizer` into separate layers again.

Reply:
This block of code could be easily moved into `Optimizer.from_params` if we just changed the call signature from `Optimizer.from_params(parameters, optimizer_params)` to `Optimizer.from_params(model, optimizer_params)`.  Since it's modifying already existing `optimizer_params` that seems like a more natural place for it.

I don't see any way to decouple completely the SLTR learning rate schedule parameters from the optimizer parameters, so suggest making it explicit and introducing a `use_slanted_triangular_rate_scheduler` or similar option in `optimizer_parameters` and only modifying the `optimizer_params["parameter_groups"]` in that case.  The code here will break / silently fail in the case where someone specifies `optimizer_params["parameter_groups"]` in their config but doesn't use the `SlantedTriangular` as it will overwrite whatever parameter group specific options they specified with the default values.  We should probably add a test to catch edge case interactions like this but that's a separate issue.

Reply:
Yes, that kind of broken interaction was what I was worried about.  Is it feasible to just require someone to specify `optimizer_params["parameter groups"]` differently in the JSON if they're using `SlantedTriangular`?  You should be able to check in the constructor of `SlantedTriangular` that things look like they're set up right, and fail if they're not.

Reply:
The point of doing it that way being that passing a parameter to `Optimizer.from_params` still gives extra logic in a `from_params` method that we want to avoid, and it still means that we're breaking the encapsulation of the learning rate scheduler and the optimizer.  Much better if the scheduler can just do what it needs to without needing to reach into the optimizer code.

Reply:
Definitely, another alternative is to require the user to specify the parameter groups themselves if they are using `SlantedTriangular`.  That will avoid this code block too, and any changes to `Optimizer.from_params`.

Reply:
Yeah, let's assume the `parameter_groups` are specified by the user as that's a lot cleaner.